addons/l10n_id_efaktur/models/account_move.py

```diff
@@ -356,16 +355,51 @@ def _generate_efaktur(self, delimiter):
         my_utf8 = output_head.encode("utf-8")
         out = base64.b64encode(my_utf8)
 
-        attachment = self.env['ir.attachment'].create({
-            'datas': out,
-            'name': 'efaktur_%s.csv' % (fields.Datetime.to_string(fields.Datetime.now()).replace(" ", "_")),
-            'type': 'binary',
-        })
+        # can only generate efaktur doc IF:
+        # (1) no linked documents for all invoices; OR
+        # (2) all invoices are linked to same document AND document strictly links to these invoices as well;
+        attachment = self.env['ir.attachment']
+        if not self.l10n_id_efaktur_document:
+            self.l10n_id_efaktur_document = self.env['l10n_id_efaktur.document'].create({'invoice_ids': self.ids})
+        elif len(self.l10n_id_efaktur_document) == 1 and set(self.l10n_id_efaktur_document.invoice_ids.ids) == set(self.ids):
+            attachment = self.l10n_id_efaktur_document.attachment_id
+        else:
+            invoices_with_document = self.filtered(lambda x: x.l10n_id_efaktur_document).mapped('name')
+            raise UserError(_("Unable to generate efaktur document on invoies with pre-existing documents: %s", ','.join(invoices_with_document)))
+
+        # create/update attachment and link it to efaktur document if new
+        if not attachment:
+            attachment = self.env['ir.attachment'].create({
+                'datas': out,
+                'name': 'efaktur_%s.csv' % (fields.Datetime.to_string(fields.Datetime.now()).replace(" ", "_")),
+                'type': 'binary',
+                'res_model': 'l10n_id_efaktur.document',
+                'res_id': self.l10n_id_efaktur_document.id,
+            })
+            self.l10n_id_efaktur_document.attachment_id = attachment.id
+        else:
+            attachment.write({
+                'datas': out,
+                'name': 'efaktur_%s.csv' % (fields.Datetime.to_string(fields.Datetime.now()).replace(" ", "_")),
+            })
 
         for record in self:
             record.message_post(attachment_ids=[attachment.id])
-        self.l10n_id_attachment_id = attachment.id
         return {
             'type': 'ir.actions.client',
             'tag': 'reload',
         }
+
+    def button_cancel(self):
+        """ When cancelling the invoice, it would unset the efaktur document for all invoices
+        involved. At the same time, also post a notification on the chatter"""
+        super().button_cancel()
+        if not self.l10n_id_efaktur_document:
+            return
+        invoices = self.l10n_id_efaktur_document.invoice_ids
+
+        invoices.l10n_id_efaktur_document = False
+        for move in invoices:
+            move.message_post(
+                body='e-Faktur document removed due to invoice cancellation',
+                subject="Unset e-Faktur document")
```

Comment:
A log is enough, no need to "post" a message.
Also please use _message_log_batch, thanks!